include/sys/dnode.h

```diff
@@ -372,6 +372,13 @@ struct dnode {
 	struct zfetch	dn_zfetch;
 };
 
+/*
+ * Since AVL already has embedded element counter, use dn_dbufs_count
+ * only for dbufs not counted there (bonus buffers) and just add them.
+ */
+#define	DN_DBUFS_COUNT(dn)	((dn)->dn_dbufs_count + \
+    avl_numnodes(&(dn)->dn_dbufs))
```

Comment:
Can we take this a step further.  Since a dnode can only ever have a single bonus buffer we should be able to entirely remove the `dn->dn_dbufs_count` member like this:

```
#define DN_DBUFS_COUNT(dn) (avl_numnodes(&(dn)->dn_dbufs) + (dn->dn_bonus != NULL))
```

Reply:
I agree with @behlendorf that further simplification would be nice, but I'm fine with the change as-is, since it's an improvement over what we have today.

If we don't have any plans to support kmem_move in Linux or FreeBSD, we could remove or ifdef-out this code entirely.

Reply:
That's fine.  If we're not confident right now about the larger change, then let's move forward with this incremental improvement.